src/compiler/checker.ts

```diff
@@ -13247,10 +13245,13 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
 
                 // Get or add a late-bound symbol for the member. This allows us to merge late-bound accessor declarations.
                 let lateSymbol = lateSymbols.get(memberName);
+                if (isFromAssignmentDeclarationMember && lateSymbol) {
+                    return links.resolvedSymbol = lateSymbol;
+                }

```

Comment:
I probably should look into allowing those from assignment declarations to merge together but tbf the behavior is not even overly consistent today:
```ts
const foo = 'foo'

class Test1 {
    constructor() {
        if (Math.random()) {
            this.prop = 12
        } else {
            this.prop = '' // ok
        }
    }
}

class Test2 {
    constructor() {
        if (Math.random()) {
            this[foo] = 12
        } else {
            this[foo] = '' // error
        }
    }
}
```